Person to card

ChangDarren opened the pull request:
Initiating  the change from Person object in AB4 to card object. The end goal is to change the list of Persons in AB4 to a list of cards. Deck structure will not be implemented in this PR. 

Not ready yet, but draft PR is called so that everyone can see and comment on how they think would be best to structure the card.

ChangDarren committed:
Create Card class

Create card class base off the person class implementation.

ChangDarren committed:
Create CardBuilder class for testing

ChangDarren committed:
Create equality test case for Card

ChangDarren committed:
Create isSameCard test case for Card

ChangDarren committed:
Create JsonAdaptedCard

ChangDarren committed:
Create UniqueCardList base on UniquePersonList

ChangDarren committed:
Merge branch 'master' into person-to-card

ChangDarren committed:
Edit formatting for Card.java

ChangDarren committed:
Edit indenting for Card.java

ChangDarren committed:
Change back wrongly edited comments and set all tabs to 4 spaces.

ChangDarren committed:
Add javadocs comment for JsonAdaptedCard

ChangDarren committed:
Fix indentation issues in TypicalCards.java

ChangDarren committed:
Create TopDeck object

ChangDarren committed:
Create VersionedTopDeck object

TopDeck with the undo feature

ChangDarren committed:
Create test cases for TopDeck

ChangDarren committed:
Fix bug in UniqueCardList

ChangDarren committed:
Alter ModelManager to fit TopDeck

Change the method signature of the Model API.

xsot commented on src/test/java/seedu/address/testutil/TypicalCards.java:
```diff
@@ -0,0 +1,29 @@
+ */
+public class TypicalCards {
+    public static final Card ADDITION = new CardBuilder().withQuestion("What is 1 + 1?")
+            .withAnswer("2").withTags("Easy").build();
```
You should avoid using "easy" as a tag as difficulty will be a special type of tag in the future.

xsot commented on src/main/java/seedu/address/model/deck/Card.java:
```diff
@@ -0,0 +1,99 @@
+
+    private final String answer;
+
+    private Set<Tag> tags = new HashSet<Tag>();
```
Why isn't this final?

xsot commented on src/main/java/seedu/address/model/deck/Card.java:
```diff
@@ -0,0 +1,99 @@
+    @Override
+    public int hashCode() {
+        int hash = 7;
+        return hash + question.hashCode() + answer.hashCode();
```
Better use the library implementation: `Objects.hash()`

xsot commented on src/main/java/seedu/address/model/deck/Card.java:
```diff
@@ -0,0 +1,99 @@
+    /**
+     * Returns true if both cards have the same question and answers.
+     * Defines a stronger notion of equality between 2 cards.
+     */
```
You forgot to check tags.

xsot commented on src/main/java/seedu/address/model/deck/Card.java:
```diff
@@ -0,0 +1,99 @@
+
+    /**
+     * Returns true if both cards have the same question.
+     * This defines a weaker notion of equality between two persons.
```
Remember to change all instances of 'person' to 'card'.

xsot commented on src/test/java/seedu/address/model/deck/CardTest.java:
```diff
@@ -0,0 +1,63 @@
+    public void isSameCard() {
+        Card additionCopy = new CardBuilder(ADDITION).build();
+
+        //same object -> returns true
```
Missing space after `//`

ChangDarren committed:
Change the JSON storage from AB4 to TopDeck

ChangDarren committed:
Edit Storage manager to change from AB4 to TopDeck

ChangDarren committed:
Change the sample data from persons to cards

ChangDarren commented:
I have changed the parts for the model and storage portion. The Logic portion of the code base however is still running base on AB4 and so the test cases will fail. Is it possible to push this as a seperate branch and then someone else help me to proceed with changing the logic section? Thank you.

ChangDarren committed:
Change LogicManager to fit TopDeck instead of AB4

ChangDarren committed:
Change AddCommand to fit TopDeck instead of AB4

ChangDarren committed:
Change Edit and Find Command to fit TopDeck instead of AB4

ChangDarren committed:
Change Redo and Select Command to fit TopDeck instead of AB4

ChangDarren committed:
Change PersonCard to CardDisplay to fit TopDeck instead of AB4

ChangDarren committed:
Change UI to fit TopDeck instead of AB4

ChangDarren committed:
Fix test utilities to match the changes from AB4 to TopDeck

ChangDarren committed:
Update the command test utility

ChangDarren committed:
Add UniqueCardList test

ChangDarren committed:
Add UniqueCardList test

ChangDarren committed:
Update ModelManager Test

ChangDarren committed:
Update JsonAdaptedCardTest

ChangDarren committed:
Update TopDeckStorageTest

ChangDarren commented:
The checks that have not been updated are:
* logic/commands
* UI

ChangDarren committed:
Fix documentation due to mass refactoring

ChangDarren committed:
Fix documentation due to mass refactoring

ChangDarren committed:
Fix documentation due to mass refactoring

ChangDarren committed:
Update Add Command tests

ChangDarren committed:
Move NameContainsKeywordsPredicate from person to deck package

ChangDarren committed:
Update Delete and Edit command test

ChangDarren committed:
Update Find, List and Select command test

ChangDarren committed:
Update Add, Edit Command parser tests

ChangDarren committed:
Update Logic Manager and Storage Manager tests

ChangDarren committed:
Updated ui tests for TopDeck

ChangDarren committed:
Update Add and Delete Command System Test

ChangDarren committed:
Update Edit and Find Command System test

ChangDarren committed:
Fix edit and find commands test

Bug found due to the difference in find for person and card. For person, first name alone should be unique but for cards, the first word may give many matches,

ChangDarren committed:
Fix Card UI tests

ChangDarren committed:
Fix Add Command System tests

ChangDarren commented:
Things left:
1. Fix the System test case for Edit, Find and Delete.
2. Replace the last remnants of AB4 (persons) to TopDeck (Card)

The UI and app is working generally, so nothing huge is broken. Yay!!

ChangDarren committed:
Update Edit Command System Test

ChangDarren committed:
Update Find Command System Test

Additions to the number of cards in the test case due to the Delete System Test.

ChangDarren committed:
Remove the * imports and get rid of unused imports

ChangDarren committed:
Remove unnecessary import statement

ChangDarren committed:
Remove unnecessary import statements and reorder the import statements by lexicographical order.

ChangDarren committed:
Reorder import statements in TopDeckTest

ChangDarren commented:
Do not merge the branch yet as there are still remnants of old code from AddressBook.

ChangDarren committed:
Remove Person and Person's test

Update ParserUtil to only parse tag
